### ISSUE TITLE
Use luajit's git history

### ISSUE THREAD

**Issue by idbrii** [iponweb/luavela#1]
Since this project is a fork of luajit, why not include their history? I [gave it a shot in a fork](https://github.com/idbrii/luavela/tree/luajit-history).

Caveats:
* It doesn't have a zero-line diff. I'm going to try to address this by doing it from a real unix machine.
* It rewrites your history (it must to add the history).
* I can't send it as a PR because I rewrote the beginning of history.
* Accepting this change won't show luavela as a "forked from LuaJIT/LuaJIT" on github.

My process

    echo Add remotes.
    git remote add luajit http://luajit.org/git/luajit-2.0.git
    git remote add vela https://github.com/iponweb/luavela.git

    read -p "Grab first commit to luavela. " pause
    git checkout -b luajit-history vela/master~

    read -p "Move head to luajit, but keep luavela changes. " pause
    git reset luajit/master
    git add -A
    git commit -C vela/master~

    read -p "Grab remaining luavela changes. " pause
    git cherry-pick vela/master 

    read -p "Diff results. " pause
    git diff --summary vela/master

I did a diff against a clean checkout of luavela and found no changes, but `git diff` finds lots of mode changes (I'm on ntfs so I can't help it) and a couple deletions that I don't understand. I can try again on a real filesystem.

**Comment by idbrii** [comment]
Briefly looking at some changes in my diff, it looks like some luajit changes aren't in your fork (LuaJIT/LuaJIT@5c911998a3c85d024a8006feafc68d0b4c962fd8
LuaJIT/LuaJIT@d4ee80342770d1281e2ce877f8ae8ab1d99e6528).

To get the most accurate history and be able to pull in similar luajit fixes, what version of luajit did you fork from?

**Comment by fsfod** [comment]
There changelog seems to only imply only up to 2.0.5 was merged which is what i used in my attempt to [compare there changes](https://github.com/fsfod/LuaJIT/commits/luavela).

**Comment by idbrii** [comment]
Updated my fork to a version with a clean diff and base it off v2.0.5.

To accept this change, you'd have to `git reset --hard idbrii/luajit-history` on your master.

Method:

    echo Add remotes.
    git remote add luajit http://luajit.org/git/luajit-2.0.git
    git remote add vela https://github.com/iponweb/luavela.git

    git fetch luajit
    git fetch vela

    read -p "Grab first commit to luavela. " pause
    git checkout -b luajit-history-again vela/master~

    read -p "Move head to luajit, but keep luavela changes. " pause
    # fsfod thinks luavela forked around v2.0.5
    git reset v2.0.5
    # Some files from luavela are in gitignore, so force.
    git add -fA
    git commit -C vela/master~

    read -p "Grab remaining luavela changes. " pause
    git cherry-pick vela/master 

    read -p "Diff results. " pause
    git diff --summary vela/master

**Comment by eliasdaler** [comment]
In theory, this sounds kind of useful - especially for doing git blame, but are there any other reasons to do it?

Starting history from LuaJIT 2.0.5 and then doing "Initial commit" doesn't show the history right - we started from LuaJIT 2.0 and then back-ported patches to our fork. And indeed, I think we're very close to LuaJIT 2.0.5 - I think almost all important patches were back-ported (we didn't back-port patches which were about processor architectures other than x86).

I think that we diverged from LuaJIT's history so far, that it's much easier to do diffs with corresponding LuaJIT files other than trying to do git blame. We renamed a lot of files, in some we completely redid the formatting... Our history should be seen as parallel to LuaJIT, not as if we started from LuaJIT 2.0.5.

We tried to preserve some of the history to show which changes to LuaJIT were ported to our fork, but later we decided not to do so, because of privacy concerns (lots of internal IPONWEB things in-between the neutral commits).

@igelhaus - any thoughts?

**Comment by idbrii** [comment]
Working on top of their history makes it easier for other people to contribute fixes from their luajit forks to luavela. They can rebase their luajit changes in top of luavela. Without luajit's history, git won't allow you to rebase. Additionally, the history makes people interested in replacing their uses of luajit with luavela able to see what's changed. (Like what fsfod is [trying to do here](https://github.com/fsfod/LuaJIT/commits/luavela).)

> it's much easier to do diffs with corresponding LuaJIT files

Without the history, that means cloning luajit, finding the corresponding file, and diffing them. It's easier to `git log --find-copies-harder --patch  filename` to see the changes -- even when the file was renamed.

For example this new history allows you can pull fixes from luajit. I did a low-effort test of pulling in changes by skipping every conflict in my [luajit-noconflict-fixes branch](https://github.com/idbrii/luavela/commits/luajit-noconflict-fixes).

Method:

```
git checkout -t luajit/master -b luajit-noconflict-fixes
git rebase luajit-history

# skip all conflicts
while true ; do
    if [[ -d ".git/rebase-merge" || -d ".git/rebase-apply" ]] ; then
        git rebase --skip
    else
        break
    fi
done
```

There's 6 changes, 4 look relevant to x86.

While starting from 2.0.5 isn't an accurate representation of how you worked, if you've backported all the changes it seems like a good representation of the final state of your code -- especially since you can't release your full history. Alternatively, [we could do the same process off of 2.0.0](https://github.com/idbrii/luavela/commits/luajit-history-v2) -- it might make it easier for other people to do their own backports of non-x86 patches, but might make things more complicated that are built on top of the changes you backported. Not sure. Doing the above process to get [luajit's fixes after v2.0.0](https://github.com/idbrii/luavela/commits/luajit-v2-noconflict-fixes) brings in 5 additional no conflict fixes.

Regardless of which version you use as a base, I think using the history would help bridge the gap between contributors/users of luajit and a new community around luavela.

**Comment by igelhaus** [comment]
Hello David @idbrii,

Sorry for not replying earlier, and many thanks for your proposal. A couple of notes for added clarity:

* We had forked long before LuaJIT 2.0.5 was released. And in fact it was not a "true" fork, at that time, we used even different VCS (Mercurial), so we simply transferred the sources to our internal infrastructure and started from there.
* However, prior to the public release we indeed made sure to backport all relevant changes up to LuaJIT 2.0.5.
* There are some commits in LuaJIT's v2.0 branch after 2.0.5 release, and we have plans to backport them soon.
* Finally, this repository is just a mirror of some internal repo managed by IPONWEB inside its infrastructure.

With all that said, I will give a try to your proposal, but given our complex internal history we have to maintain, some infrastructure concerns and (alas) my personal lack of expertise with git I cannot give any hard promises. If something goes wrong, we'll leave it as is.

I will update this issue with the results of the experiment.
